packages/doenetml-worker-rust/doenetml-core/src/state_var_calculations.rs

```diff
@@ -0,0 +1,762 @@
+    instruction_idx: usize,
+    val_idx: usize,
+    dependency_instructions: Option<Vec<DependencyInstruction>>,
+    instruct_dependencies: Option<Vec<Dependency>>,
+    dependencies_for_state_var: Option<Vec<Vec<Dependency>>>,
+}
+
+/// The state needed to save the partial progress of a state variable that is resolved but still stale.
+///
+/// Since it is resolved, all dependencies have been calculated, so we just need to keep track
+/// of which instruction and value from the dependencies we are freshening.
+#[derive(Debug)]
+pub struct StaleCalculationState {
+    component_state: ComponentStateDescription,
+    instruction_idx: usize,
+    val_idx: usize,
+}
+
+/// Freshen all the state variables for a component that are designated as rendered
+/// and recurse to rendered children.
+///
+/// Returns a vector of the indices of the components reached.
+pub fn freshen_all_stale_renderer_states(
+    stale_renderers: &mut Vec<ComponentIdx>,
+    components: &Vec<Rc<RefCell<ComponentEnum>>>,
+    dependencies: &mut Vec<Vec<Vec<Vec<Dependency>>>>,
+    dependent_on_state_var: &mut Vec<Vec<Vec<ComponentStateDescription>>>,
+    dependent_on_essential: &mut Vec<HashMap<EssentialDataOrigin, Vec<ComponentStateDescription>>>,
+    essential_data: &mut Vec<HashMap<EssentialDataOrigin, EssentialStateVar>>,
+    freshen_stack: &mut Vec<StateVarCalculationState>,
+    should_initialize_essential_data: bool,
+) -> Vec<usize> {
+    // recursively get a list of all rendered descendants of the components in stale_renderers
+    let mut stale_renderer_idx = 0;
+    while stale_renderer_idx < stale_renderers.len() {
+        let component_idx = stale_renderers[stale_renderer_idx];
+        stale_renderer_idx += 1;
+
+        stale_renderers.extend(get_non_string_rendered_children_including_from_extend(
+            component_idx,
+            components,
+        ));
+    }
+
+    // deduplicate the list of stale renderers,
+    // sorting first since .dedup removes only consecutive duplicates.
+    stale_renderers.sort_unstable();
+    stale_renderers.dedup();
+
+    for component_idx in stale_renderers.iter() {
+        let rendered_state_var_indices = components[*component_idx]
+            .borrow()
+            .get_rendered_state_variable_indices()
+            .clone();
+
+        for state_var_idx in rendered_state_var_indices {
+            let component_state = ComponentStateDescription {
+                component_idx: *component_idx,
+                state_var_idx,
+            };
+            freshen_state_var(
+                component_state,
+                components,
+                dependencies,
+                dependent_on_state_var,
+                dependent_on_essential,
+                essential_data,
+                freshen_stack,
+                should_initialize_essential_data,
+            );
+        }
+    }
+
+    let components_freshened = stale_renderers.clone();
+
+    stale_renderers.clear();
```

Comment:
The only reason for keeping `stale_renderers` around and just clearing it is to keep the space allocated for the next time this function is called. Since this function is going to be called repeated, I was trying to optimize the number of allocations needed. (Though, unlike the `freshen_stack`, each entry is only an integer.)

Reply:
If you clear this, I think it should be done at the call point rather than in this function, and it should be given a name that makes it clear that it is a scratch vector. Though, it sounds like a premature optimization to me...

Reply:
Actually, I had it wrong how this works. The `DoenetMLCore.stale_renderers` is pushed to whenever a rendered state variable is marked stale. So it needs to be cleared at the end to indicate that there are no more stale renderers. (It's not just a temporary data structure like the stacks, like I erroneously implied.)

And, by looking at this I found a flaw where `stale_renderers` gets pushed to whenever any component has a `for_renderer` state variable marked stale. Instead it should be pushed to only when a **rendered** component has a `for_renderer` state variable marked stale. Will see if this is an easy fix or an issue to be created for later. I don't yet have a quick way to tell if a component is rendered.